[[WIP]] [Issue: 103] Proof point: See if notebook-http mode can be made a "plugin"

nitind opened the pull request:
NOT READY TO BE MERGED

Relocates notebook-http mode and jupyter-websocket setup into separate modules.
For https://github.com/jupyter/kernel_gateway/issues/103.
(c) Copyright IBM Corp. 2016

parente commented:
Nice first step.

Two thoughts as you continue ...

1. I wonder if it will be possible to totally separate the modules required for the two modes (and other future modes) from each other into packages. It looks like right now there are a few things shared. Mixins and base handlers are examples of things that probably should be shared and live in KG "core". The kernel pool concept, on the other hand, only came about to support notebook-http mode. The degenerate one that exists for jupyter-websocket mode is only there for API parity in a sense. That's a case where it might be possible to have that module only in the notebook mode package and remove the stubbed one from being used in the websocket package.

2. Originally we put everything into a `services` directory to mirror the structure of the notebook project. With the idea of plugging in your own personality, we'll need to break with having one big services directory. Maybe `services/` becomes the place where common code across personalities lives, while `plugins` or `personalities` or ... becomes the directory where the implementations that ship with KG live. How the `--KernelGatewayApp.api=a_personality_package` gets implemented later for the issue will dictate part of this.

parente commented:
In passing while working on other issues, I noted that even though there's not an explicit mode check in our custom kernel manager, there's definitely still notebook-http specific code like the use of APICellParser here https://github.com/jupyter/kernel_gateway/blob/master/kernel_gateway/services/kernels/manager.py#L85. This should be refactored as well. But we need to be careful: the logic of pre-populating kernel memory is also used in jupyter-websocket mode. It just has special case logic for executing only some cells in notebook-http mode.

Maybe the kernel manager also needs a delegation model? Or maybe each package provides its own kernel manager? How much code is redundant?

nitind commented:
The degenerate kernel pool would seem expendable, but that's still used for the prespawning of kernels in jupyter-websocket mode. The question of whether the tornado loop is itself something to break out is probably not in scope.

For the APIParser, it would be simpler to let the provide the kernel manager with a test function for `filter` determining which cells to seed.

I'm not married to the name "personality", btw.

nitind commented:
What's in there now:
- changed "API" parameter from a string to a fully qualified module name, with the module expected to implement a `create_personality` factory function (the previous `jupyter-websocket` and `notebook-http` values will continue to work along side the module names)
- moved the current `jupyter-websocket` and `notebook-http` functionality into their own modules, no longer buried in `services/`
- relocate functionality specific to either mode into their modules, like the API cell parser and swagger generation for `notebook-http` mode

Relocating traitlets unique to one personality will require more exploration in its own PR, assuming we figure out how to do it.

parente commented on kernel_gateway/gatewayapp.py:
```diff
@@ -221,6 +211,21 @@ def _api_changed(self, name, old, new):
+        if module_name == 'jupyter-websocket':
+            module_name = 'kernel_gateway.jupyter_websocket'
+        elif module_name == 'notebook-http':
+            module_name = 'kernel_gateway.notebook_http'
```
Cool. Glad this was straightforward.

parente commented on kernel_gateway/gatewayapp.py:
```diff
@@ -221,6 +211,21 @@ def _api_changed(self, name, old, new):
+            api_module = importlib.import_module(module_name)
+        except ImportError:
+            # a ValueError will be triggered in the caller
+            pass
```
Better to let the ImportError bubble. It's possible the user has provided a valid package name, but that package cannot be imported for other reasons (e.g., it is throwing an exception because of a missing dependency).

nitind replied:
Good point.

parente commented on kernel_gateway/gatewayapp.py:
```diff
@@ -312,19 +320,23 @@ def init_configurables(self):
+        func = getattr(api_module, 'create_personality', None)
+        if func is None:
+            self.log.critical('ERROR: the personality factory function was not found in the API module {}'.format(self.api))
+            self.exit(1)
```
Probably better to raise a RuntimeError here with the message you're logging as critical. `sys.exit(1)` is harsh when testing.

parente commented on kernel_gateway/gatewayapp.py:
```diff
@@ -312,19 +320,23 @@ def init_configurables(self):
+            'prespawn_count': self.prespawn_count,
+            'kernel_manager': self.kernel_manager,
+            'base_url': self.base_url,
+            'seed_uri': self.seed_uri})
```
These might build up over time. How about just passing `self` and letting the personality access whatever it needs from the application instance?

parente commented on kernel_gateway/jupyter_websocket/__init__.py:
```diff
@@ -0,0 +1,65 @@
+from ..base.handlers import default_handlers as default_base_handlers
```
Copy/paste the Jupyter copyright header to the top of the file from one of the others.

parente commented on kernel_gateway/gatewayapp.py:
```diff
@@ -312,19 +320,23 @@ def init_configurables(self):
-            )
+        
+        api_module = self._load_api_module(self.api)
+        func = getattr(api_module, 'create_personality', None)
```
I'm a little confused about the second arg, the `None`. When is this not None? Why pass it in if it's always None?

nitind replied:
It's the default return value if not found, not an arg being passed during invocation. Like the possible ImportError above, it's simpler to let it fail as an AttributeError with the reported line showing what was not found than to intentionally do anything more elaborate...right?

parente replied:
That's what happens when I do code review at night. Somehow I missed the `getattr` and read it as invoking `creating_personality`. 👀 

Yes. I'd let the AttributeError bubble. If they've passed a valid package that doesn't have that function, let's fail fast and ugly.

parente commented on kernel_gateway/jupyter_websocket/__init__.py:
```diff
@@ -0,0 +1,65 @@
+    '''Personality for standard websocket functionality, registering
+    endpoints that are part of the Jupyter Kernel Gateway API
+    '''
+    def __init__(self, kernelspec):
```
It seems odd to have `kernelspec` be the one and only constructor arg. I think I know why you're passing it for notebook http server, but I only see `None` ever passed when creating a personality, so maybe I'm missing how it's intended used. 

Still, it seems like there should be a more elegant way. Would passing the gatewayapp instance to the constructor and accessing its attributes help?

nitind replied:
It can be used to solve the same problem cropping up during tests, yes. Even waiting to instantiate the SessionManager until *after* calling `self.personality.init_configurables()` still leaves a race condition in the tests:
```
tornado.application: ERROR: Future <tornado.concurrent.Future object at 0x7feb145ebb00> exception was never retrieved: Traceback (most recent call last):
  File "/opt/conda/lib/python3.5/site-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/srv/kernel_gateway/kernel_gateway/services/kernels/manager.py", line 82, in start_kernel
    if self.parent.personality.should_seed_cell(self.seed_kernelspec, code):
  File "/srv/kernel_gateway/kernel_gateway/notebook_http/__init__.py", line 84, in should_seed_cell
    return (not self.api_parser.is_api_cell(code) and not self.api_parser.is_api_response_cell(code))
AttributeError: 'NotebookHTTPPersonality' object has no attribute 'api_parser'
```

parente commented on kernel_gateway/tests/services/cell/test_parser.py:
```diff
@@ -5,7 +5,7 @@
 import unittest
 import re
 import sys
-from kernel_gateway.services.cell.parser import APICellParser
```
We should refactor the test folders to following the renaming of the services folders? For example, this one should be in `tests/notebook_http/cell/test_parser.py` probably.

parente commented on kernel_gateway/notebook_http/__init__.py:
```diff
@@ -0,0 +1,89 @@
+from ..base.handlers import default_handlers as default_base_handlers
```
Copy/paste the Jupyter copyright header to the top of the file from one of the others.

parente commented on kernel_gateway/jupyter_websocket/__init__.py:
```diff
@@ -0,0 +1,65 @@
+        self.kernel_pool.shutdown()
+
+def create_personality(gatewayapp, kernel_name):
+    return JupyterWebsocketPersonality(kernel_name)
```
Where is `kernel_name` set to a non-`None` value?

nitind replied:
JupyterWebsocketPersonality doesn't make use of it directly. The kernel manager uses the value from the gatewayapp itself.

parente commented:
This is looking good! 🎉 

I left comments inline. Some simple fixes, a couple stylistic things, and one part I'm confused about (the kernel_name arg). I think we need to clean-up the personality API a bit (e.g., only pass gateway app to the constructor, let the instance store it from there and use it in other methods). Other than that, I think it's nearly ready to go in.

I'll wait for you to comment back before I pull and try here.

nitind commented:
Closing in favor of cleaner merge from PR https://github.com/jupyter/kernel_gateway/pull/171 .
